robot_framework/process.py

```diff
@@ -1,8 +1,204 @@
+
+    # Prepare access to service platform
+    kombit_access = KombitAccess(service_cvr, certificate_dir, True)
+
+    # Prepare access to email
+    graph_credentials = orchestrator_connection.get_credential(config.GRAPH_API)
+    graph_access = graph_authentication.authorize_by_username_password(graph_credentials.username, **json.loads(graph_credentials.password))
+    mails = graph_mail.get_emails_from_folder(config.EMAIL_USER, config.MAIL_SOURCE_FOLDER, graph_access)
+
+    for mail in mails:
+        orchestrator_connection.log_trace("Reading emails.")
+        # Get attachment from email
+        attachments = graph_mail.list_email_attachments(mail, graph_access)
+        email_attachment = graph_mail.get_attachment_data(attachments[0], graph_access)
+        # Get data from email text
+        requester = _get_recipient_from_email(mail.body)
+        request_type = _get_request_type_from_email(mail.body)
+        # Send and delete email
+        start_time = time.time()
+        return_data, rows_handled = handle_data(email_attachment, kombit_access, request_type, thread_count)
+
+        orchestrator_connection.log_info(f"{rows_handled} Rows handled. Total time spent: {time.time()-start_time} seconds")
+        _send_status_email(requester, return_data)
+        graph_mail.delete_email(mail, graph_access)
+
+
+def handle_data(input_file: BytesIO, access: KombitAccess, service_type: Literal['Digital Post', 'NemSMS', 'Begge'], thread_count: int) -> BytesIO:
+    """ Read data from attachment, lookup each CPR number found and return a new file with added data.
+
+    Args:
+        input_file: Excel-file with rows of CPR to work on
+        access: Kombit Access Token
+        service_type: 'digitalpost', 'nemsms' or 'begge', to set which lookups to perform
+    Returns:
+        Filtered and formatted file with a list of people indicating whether they have Digital Post or not.
+    """
+    workbook = load_workbook(input_file)
+    input_sheet: Worksheet = workbook.active
+
+    # Check which services are requested
+    service = ["Digital Post", "NemSMS"] if service_type == "Begge" else [service_type]
+
+    # Call digital_post.is_registered for each input row and each required service
+    data = threaded_service_check(input_sheet, service, access, thread_count)
+    # data = linear_service_check(input_sheet, service, access)
+
+    # Add data to excel sheet
+    write_data_to_output_excel(service, data, input_sheet)
+
+    # Grab workbook from memory and return it
+    byte_stream = BytesIO()
+    workbook.save(byte_stream)
+    byte_stream.seek(0)
+    return byte_stream, input_sheet.max_row
+
+
+def threaded_service_check(input_sheet: Worksheet, service: List[str], kombit_access: KombitAccess, thread_count: int) -> dict[str, dict[str, bool]]:
+    """ Call digital_post.is_registered for each input row and each required service.
+
+    Args:
+        input_sheet (Worksheet): The input worksheet containing rows of data.
```

Comment:
We normally don't have the types in the docstring when it's already in the def :)